src/middleware/compression.rs

```diff
@@ -0,0 +1,271 @@
+    Context, Error, Response,
+};
+
+lazy_static! {
+    static ref RE_ACCEPT_ENCODING: Regex =
+        Regex::new(r"(?P<a>[\w\*]+)(;q=(?P<q>\d+(\.\d{1,2})?))?(\s*,\s*)?").unwrap();
+    static ref RE_CONTENT_ENCODING: Regex = Regex::new(r"(?P<a>[\w\*]+)(\s*,\s*)?").unwrap();
+}
+
+/// Encodings for Compression/Decompression middleware.
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub enum Encoding {
+    Gzip,
+    Deflate,
+    Identity,
+    Br,
+}
+
+impl Encoding {
+    fn header_value(self) -> HeaderValue {
+        match self {
+            Encoding::Gzip => HeaderValue::from_str("gzip").unwrap(),
+            Encoding::Deflate => HeaderValue::from_str("deflate").unwrap(),
+            Encoding::Identity => HeaderValue::from_str("identity").unwrap(),
+            Encoding::Br => HeaderValue::from_str("br").unwrap(),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct EncodeSettings {
+    gzip: gzip::Compression,
+    deflate: zlib::Compression,
+    brotli: brotli::CompressParams,
+}
+
+impl Default for EncodeSettings {
+    fn default() -> Self {
+        Self {
+            gzip: Default::default(),
+            deflate: Default::default(),
+            brotli: brotli::CompressParams::new(),
+        }
+    }
+}
+
+/// Middleware for handling outgoing response compression.
+#[derive(Debug)]
+pub struct Compression {
+    default_encoding: Encoding,
+    settings: EncodeSettings,
+}
+
+impl Default for Compression {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Compression {
+    pub fn new() -> Self {
+        Self {
+            default_encoding: Encoding::Identity,
+            settings: Default::default(),
+        }
+    }
+
+    pub fn with_default(default_encoding: Encoding) -> Self {
+        Self {
+            default_encoding,
+            settings: Default::default(),
+        }
+    }
+
+    fn parse_encoding(&self, s: &str) -> Result<Encoding, Error> {
+        match s {
+            "gzip" => Ok(Encoding::Gzip),
+            "deflate" => Ok(Encoding::Deflate),
+            "identity" => Ok(Encoding::Identity),
+            "br" => Ok(Encoding::Deflate),
+            "*" => Ok(self.default_encoding),
+            _ => Err(Error::from(StatusCode::UNSUPPORTED_MEDIA_TYPE)),
+        }
+    }
+
+    fn preferred_encoding(&self, headers: &HeaderMap) -> Encoding {
+        // TODO: test on invalid Accept-Encoding values
+        let mut preferred_encoding = self.default_encoding;
+        let mut max_qval = 0.0;
+        'outer: for hval in headers.get_all(ACCEPT_ENCODING).iter() {
+            for cap in RE_ACCEPT_ENCODING.captures_iter(hval.to_str().unwrap()) {
+                match self.parse_encoding(cap.name("a").unwrap().as_str()) {
+                    Ok(encoding) => match cap.name("q") {
+                        Some(qval) => {
+                            let qval = match qval.as_str().parse::<f32>() {
+                                Ok(f) => f,
+                                Err(_) => {
```

Comment:
```suggestion
                                Err(_) => continue,
```
Why not just inline it?

Reply:
Oh, good point. I'll also fix that along with the other review you left below.